tests/multiple_database/tests.py

```
@@ -1502,18 +1498,15 @@ def test_database_routing(self):
         self.assertEqual(Book.objects.using("default").count(), 1)
         self.assertEqual(Book.objects.using("other").count(), 1)
 
-    def test_invalid_set_foreign_key_assignment(self):
```

Instead of deleting the test, we should update the test to show the error is not required because this works. I would expect something like:
```python
    def test_set_foreign_key_assignment(self):
        marty = Person.objects.using("default").create(name="Marty Alchin")
        dive = Book.objects.using("other").create(
            title="Dive into Python",
            published=datetime.date(2009, 5, 4),
        )
        marty.edited.set([dive])
        self.assertSequenceEqual(marty.edited.all(), [dive])
```

However, this fails. Can you look into this please?

---

That the error `"<Book: Dive into Python> instance isn't saved. Use bulk=False or save the object first."` is gone is an issue to me. This feels like a silent failure.
I used to have a helpful error telling me what I should do instead :thinking:

---

> That the error "<Book: Dive into Python> instance isn't saved. Use bulk=False or save the object first." is gone is an issue to me. This feels like a silent failure.

Having removed the requirement of `_state.db` equality between related objects, when we try to assign an unsaved instance we get the less specific  `TypeError` with message "Model instances without primary key value are unhashable" (not a silent failure but indeed in comparison it feels like one ,-).

Aktshyually, I reformulated the test as `test_invalid_set_unsaved_assignment` to test that an unsaved asignment still failed:

```python
    def test_invalid_set_unsaved_assignment(self):
        marty = Person.objects.using("default").create(name="Marty Alchin")
        dive = Book(
            title="Dive into Python",
            published=datetime.date(2009, 5, 4),
        )
        # Set a foreign key set with an object that has not been saved
        msg = "Model instances without primary key value are unhashable"
        with self.assertRaisesMessage(TypeError, msg):
            marty.edited.set([dive])
```

> I used to have a helpful error telling me what I should do instead 

I agree that the previous error message is more helpful.

Let me see if that behavior can be reinstated ...

---

I'm not sure if I was clear, in the original test `test_invalid_set_foreign_key_assignment`
```python
        marty = Person.objects.using("default").create(name="Marty Alchin")
        dive = Book.objects.using("other").create(
            title="Dive into Python",
            published=datetime.date(2009, 5, 4),
        )
        marty.edited.set([dive])  # This used to error... I think this is "failing silently"
```

---

Yes, this used to be an error.

And it still is when the database router does not allow two establish a relation between two instances of different databases, as in `test_m2m_cross_database_protection`:

https://github.com/django/django/blob/0bac41fc7e4a842e8d20319cba31cc645501c245/tests/multiple_database/tests.py#L435-L443

It should however not be an error to relate an instance fetched from a read-only connection to an instance updated using a read-write connection, both to the same database cluster.

Unfortunately because a requirement was about the connections used to create or fetch the related instances. In the example of the `test_invalid_set_foreign_key_assignment` the `dive._state.db` was not equal `marty._state.db` and an error would be raised ... despite the `TestRouter` explicitely mimics connections to replica and allowing a relation between two objects instanciated with different connections:

```python
    def allow_relation(self, obj1, obj2, **hints):
        return obj1._state.db in ("default", "other") and obj2._state.db in (
            "default",
            "other",
        )
```

So the requirement about `_state.db` made here:

https://github.com/django/django/blob/0bac41fc7e4a842e8d20319cba31cc645501c245/django/db/models/fields/related_descriptors.py#L784-L792

and here:

https://github.com/django/django/blob/0bac41fc7e4a842e8d20319cba31cc645501c245/django/contrib/contenttypes/fields.py#L682-L689

both defeat the purpose of the database router's `allow_relation` method when relating objects.

It may be possible to circumvent this issue by using the `bulk=False` flag but this is not an option when using third party software that do not, like the Django REST framework for instance.

That is why I propose here to remove the conditions of equality between `_state.db` and let the database router's `allow_relation` decide if two instances from different database *connections* can be related.

A side-effect of this fix is to change the type and label of the exception raised when trying to set or add a relation to an unsaved instance.